kafka-ballerina/tests/producer_client_tests.bal

```diff
@@ -0,0 +1,170 @@
+    result = closeTestProducer->close();
+    test:assertFalse(result is error, result is error ? result.toString() : result.toString());
+    result = closeTestProducer->send({ topic: topic3, value: message.toBytes() });
+    test:assertTrue(result is error);
+    error receivedErr = <error>result;
+    string expectedErr = "Failed to send data to Kafka server: Cannot perform operation after producer has been closed";
+    test:assertEquals(receivedErr.message(), expectedErr);
+}
+
+@test:Config {}
+function producerFlushTest() returns error? {
+    Producer flushTestProducer = check new (DEFAULT_URL, producerConfiguration);
+    check flushTestProducer->send({ topic: topic1, value: TEST_MESSAGE.toBytes() });
+    check flushTestProducer->'flush();
+    ConsumerConfiguration consumerConfiguration = {
+        topics: [topic1],
+        offsetReset: OFFSET_RESET_EARLIEST,
+        groupId: "producer-flush-test-group",
+        clientId: "test-producer-05"
+    };
+    Consumer stringConsumer = check new (DEFAULT_URL, consumerConfiguration);
+    ConsumerRecord[] consumerRecords = check stringConsumer->poll(3);
+    test:assertEquals('string:fromBytes(consumerRecords[0].value), TEST_MESSAGE);
+}
+
+@test:Config {}
+function producerGetTopicPartitionsTest() returns error? {
+    Producer topicPartitionTestProducer = check new (DEFAULT_URL, producerConfiguration);
+    TopicPartition[] topicPartitions = check topicPartitionTestProducer->getTopicPartitions(topic1);
+    test:assertEquals(topicPartitions[0].partition, 0, "Expected: 0. Received: " + topicPartitions[0].partition.toString());
+    check topicPartitionTestProducer->close();
+}
+
+@test:Config {}
+function transactionalProducerTest() returns error? {
+    ProducerConfiguration producerConfigs = {
+        clientId: "test-producer-06",
+        acks: "all",
+        retryCount: 3,
+        enableIdempotence: true,
+        transactionalId: "test-transactional-id"
+    };
+    Producer transactionalProducer = check new (DEFAULT_URL, producerConfigs);
+    transaction {
+        check transactionalProducer->send({
+            topic: topic3,
+            value: TEST_MESSAGE.toBytes(),
+            partition: 0
+        });
+        var commitResult = commit;
+        if (commitResult is ()) {
+            io:println("Commit successful");
+        } else {
+            test:assertFail(msg = "Commit Failed");
+        }
+    }
+    ConsumerConfiguration consumerConfiguration = {
+        topics: [topic3],
+        offsetReset: OFFSET_RESET_EARLIEST,
+        groupId: "consumer-transactional-test-group",
+        clientId: "test-consumer-38"
+    };
+    Consumer consumer = check new (DEFAULT_URL, consumerConfiguration);
+    ConsumerRecord[] consumerRecords = check consumer->poll(5);
+    test:assertEquals(consumerRecords.length(), 4, "Expected: 4. Received: " + consumerRecords.length().toString());
```

Comment:
Not sure how you got 4 records here by producing once. Could be due to reusing topics in multiple tests?

Reply:
Yes. Ill make the topics independent.